swiftide-integrations/src/pgvector/mod.rs

```diff
@@ -0,0 +1,163 @@
+//! This module integrates with the pgvector database, providing functionalities to create and manage vector collections,
+//! store data, and optimize indexing for efficient searches.
+//!
+//! pgvector is utilized in both the `indexing::Pipeline` and `query::Pipeline` modules.
+mod persist;
+mod pgv_table_types;
+use anyhow::Result;
+use derive_builder::Builder;
+use sqlx::PgPool;
+use std::fmt;
+
+use pgv_table_types::{FieldConfig, MetadataConfig, PgDBConnectionPool, VectorConfig};
+
+const DEFAULT_BATCH_SIZE: usize = 50;
+
+/// Represents a Pgvector client with configuration options.
+///
+/// This struct is used to interact with the Pgvector vector database, providing methods to manage vector collections,
+/// store data, and ensure efficient searches. The client can be cloned with low cost as it shares connections.
+#[derive(Builder, Clone)]
+#[builder(setter(into, strip_option), build_fn(error = "anyhow::Error"))]
+pub struct PgVector {
+    /// Database connection pool.
+    #[builder(default = "PgDBConnectionPool::default()")]
+    connection_pool: PgDBConnectionPool,
+
+    /// Table name to store vectors in.
+    #[builder(default = "String::from(\"swiftide_pgv_store\")")]
+    table_name: String,
+
+    /// Default sizes of vectors. Vectors can also be of different
+    /// sizes by specifying the size in the vector configuration.
+    vector_size: Option<i32>,
```

Comment:
This value isn't optional, is it? What happens if it is None?

Reply:
Just to clarify: this parameter can’t be `None` because if it is, the user won’t be able to build the `PgVector::fields` parameter, and `PgVectorBuilder::with_vector()` would fail. We’re ensuring that users configure this parameter correctly before launching the indexing pipeline.

I’d love to hear your thoughts and any suggestions for enhancing this approach.

Reply:
Why make it an Option then? On the builder side it's still an option, when build, It shouldn't be.

Reply:
Thank you for this review feedback!

You made an excellent point about the `vector_size` and `batch_size` fields not requiring the `Option<T>` type. I've updated the PgVector implementation accordingly:

```rust
pub struct PgVector {
    /// Default vector size; can be customized per configuration.
    vector_size: i32,

    /// Batch size for storing nodes.
    #[builder(default = "BATCH_SIZE")]
    batch_size: usize,
}
```

I missed that detail when working on the `PgVectorBuilder` build function. I've completed the refactoring and cleanup.